_posts/2020-03-16-dask-usage-at-blue-yonder.markdown

```diff
@@ -1,131 +1,131 @@
+  But it does not stop there. For performance reasons, we install the Python packages holding the code needed
+  for the computations on each worker. Since the different components of our compute pipeline are implemented
+  and maintained as completely separate software packages whose requirements are not synchronized, this also
+  means we have to run a separate Dask cluster for each of the components. This leads to us operating more than
+  ten Dask clusters per customer and environment, with most of the time, only one of the clusters being active
+  and computing something. While this leads to overhead in terms of administration and hardware resouces,
+  it also gives us a lot of flexibility. For instance, we can update the software on the cluster of one part of the compute pipeline
+  while another part of the pipeline is computing something on a different cluster.
+
+  ### Some numbers
+
+  The number and size of the workers varies from cluster to cluster depending on the degree of parallelism
+  of the computation being performed, its resource requirements, and the available timeframe for the computation.
+  At the time of writing this, we are running more than 500 distinct clusters.
+  They have between one and 225 workers, with worker size varying between 1GiB and 64GiB of memory.
+
+  ## Cluster scaling and resilience
+
+  To improve manageability, resilience, and resource utilization, we run the Dask clusters on top
+  of [Apache Mesos](http://mesos.apache.org/)/[Aurora](http://aurora.apache.org/). This
+  means it is really easy to commission or decommission Dask clusters and to add or remove workers to/from
+  an existing Dask cluster. At the moment, we are starting a migration from Mesos/Aurora towards 
+  [Kubernetes](https://kubernetes.io/)
+
+  Running on top of a system like Mesos or Kubernetes provides us with resilience since a failing worker 
+  (for instance, as result of a failing hardware node)
+  can simply be restarted on another node of the system. 
+
+  Running 500 Dask clusters requires a lot of hardware. We have put two measures in place to improve
+  the utilization of hardware resources: oversubscription and autoscaling.
+
+  ### Oversubscription
+
+  [Oversubscription](http://mesos.apache.org/documentation/latest/oversubscription/) is a feature of Mesos
+  that allows to allocate more resources than physically present to services running on the system.
+  This is based on the assumption that not all services exhaust all of their allocated resources at the same time.
+  If the assumption is violated, classifying services into tiers allows to preempt less important services
+  in favor of critical ones.
+  We use this to re-purpose the resources allocated for production clusters but not utilized the whole time
+  and use them for development and staging systems.
+
+  ### Autoscaling
+
+  Autoscaling is a mechanism we implemented to dynamically adapt the number of workers in a Dask cluster
```

Comment:
Do we work on an assumption people know what autoscaling is? I don't know if it is common lingo in the data-engg world.

Reply:
The first sentence is intended as a definition of the term, as it is not common, but was created by us.